### PR TITLE
[SPARK-23774][SQL] `Cast` to CHAR/VARCHAR should truncate the values

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR aims to fix the following `CAST` behavior on `CHAR/VARCHAR` types in order to be consistent with Hive. Since `HiveStringType` is used only in parsing, this PR is also handling these cases at parsing phase.

**Spark (2.1.2 ~ 2.3.1)**
```scala
scala> sql("SELECT CAST('123' AS CHAR(1)), CAST('123' AS VARCHAR(1))").show
+-------------------+-------------------+
|CAST(123 AS STRING)|CAST(123 AS STRING)|
+-------------------+-------------------+
|                123|                123|
+-------------------+-------------------+

scala> sql("SELECT CAST('123' AS CHAR(0)), CAST('123' AS VARCHAR(0))").show
+-------------------+-------------------+
|CAST(123 AS STRING)|CAST(123 AS STRING)|
+-------------------+-------------------+
|                123|                123|
+-------------------+-------------------+
```

**Hive (1.2.1 ~ 2.3.3)**

```
hive> SELECT CAST('123' AS CHAR(1)), CAST('123' AS VARCHAR(1));
OK
1	1

hive> SELECT CAST('123' AS CHAR(0)), CAST('123' AS VARCHAR(0));
FAILED: RuntimeException Char length 0 out of allowed range [1, 255]
```

## How was this patch tested?

Pass the Jenkins with newly added test cases and existing one

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -1152,7 +1152,18 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a [[Cast]] expression.
    */
   override def visitCast(ctx: CastContext): Expression = withOrigin(ctx) {
-    Cast(expression(ctx.expression), visitSparkDataType(ctx.dataType))
+    typedVisit[DataType](ctx.dataType) match {
+      case t: CharType =>
+        validate(t.length > 0, s"Char length ${t.length} is out of range [1, 255]", ctx)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: This one and line 1161 is the same error message with Hive.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile .
Could you review this PR?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -2792,4 +2793,31 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("`Cast` to CHAR/VARCHAR should truncate the values") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Also, we need tests in `ExpressionParserSuite`?

**Comment by maropu** [comment]
Since this is a behaviour change, I think we need to update the migration guide?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -1152,7 +1152,18 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a [[Cast]] expression.
    */
   override def visitCast(ctx: CastContext): Expression = withOrigin(ctx) {
-    Cast(expression(ctx.expression), visitSparkDataType(ctx.dataType))
+    typedVisit[DataType](ctx.dataType) match {
+      case t: CharType =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I do not think this is the right thing we should do unless we can natively support CHAR and VARCHAR.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2792,4 +2793,40 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+
+      checkAnswer(
+        sql("SELECT CAST(a AS CHAR(2)), CAST(a AS CHAR(3)), CAST(a AS CHAR(4)) FROM t"),
+        Row("ab", "abc", "abc") :: Row(null, null, null) :: Nil)
+
+      sql(
+        """
+          |CREATE TABLE t_ctas
+          |USING ORC
+          |AS SELECT CAST(a AS CHAR(2)) c1, CAST(a AS CHAR(3)) c2, CAST(a AS CHAR(4)) c3 FROM t
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I agree that this is partial. Actually, this CTAS test case was I'd like to support in this PR.
Do you think Spark can have `CHAR`/`VARCHAR` in 2.4, @gatorsmile ?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Without native support of new data type CHAR and VARCHAR, supporting the syntactic sugar is misleading to end users, in my opinion.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Currently, this PR only resolves the top -level CHAR and VARCHAR. When they are embedded, does it still work? To me, the whole solution is hacky.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: What do you mean by `embedded`? This works in a subquery and in an expression like function arguments. Could you give us some examples?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: inside another expression.

**Comment by dongjoon-hyun** [comment]
Based on the review opinion, I'll close this PR and JIRA issue for now.

**Pull request closed without merging**